Jpeg encoder complete rewrite

br3aker opened the pull request:
### Prerequisites

- [x] I have written a descriptive pull-request title
- [x] I have verified that there are no overlapping [pull-requests](https://github.com/SixLabors/ImageSharp/pulls) open
- [x] I have verified that I am following the existing coding patterns and practice as demonstrated in the repository. These follow strict Stylecop rules :cop:.
- [x] I have provided test coverage for my change (where applicable)

### New architecture
Jpeg encoder now have a brand new architecture which allows us to define encoding 'configs' like this:

```cs
// YCbCr 4:1:0
new JpegFrameConfig(
    JpegColorSpace.YCbCr,
    JpegEncodingColor.YCbCrRatio410,
    new JpegComponentConfig[]
    {
        new JpegComponentConfig(id: 1, hsf: 4, vsf: 2, quantIndex: 0, dcIndex: 0, acIndex: 0),
        new JpegComponentConfig(id: 2, hsf: 1, vsf: 1, quantIndex: 1, dcIndex: 1, acIndex: 1),
        new JpegComponentConfig(id: 3, hsf: 1, vsf: 1, quantIndex: 1, dcIndex: 1, acIndex: 1),
    },
    yCbCrHuffmanConfigs,
    yCbCrQuantTableConfigs)
```

Which allow us to define any allowed subsampling setup or even a 'custom' scenario with any number of components if user needs to. This API is internal for now - user can only choose encoding mode via `enum` which we had a long time ago. Maybe one day we can make it public - who knows.

Due to this change: https://github.com/SixLabors/ImageSharp/issues/1713 is resolved and https://github.com/SixLabors/ImageSharp/issues/1476 is resolved.

It would also be fairly easy to implement optimized huffman tables or progressive encoding with current architecture.

### API changes

1. All of these encoding modes are implemented now:

```cs
YCbCrRatio420 = 0,
YCbCrRatio444 = 1,
YCbCrRatio422 = 2,
YCbCrRatio411 = 3,
YCbCrRatio410 = 4,
Luminance = 5,
Rgb = 6,
Cmyk = 7,
Ycck = 8,
```

2. Grayscale encoding now has a vectorized color conversion which gained a lot of performance.

~~3. I've removed `Ycck` encoding simply because it's not popular and we didn't support it anyway (we still can decode `YccK` though). This issue is partially done: https://github.com/SixLabors/ImageSharp/issues/808~~ 

4. Added non-interleaved encoding for any color type.

### Performance

Encoder got a bit slower but I think new architecture's worth it - we were really fast before, now we are still fast enough.

Main branch

|    Method | TargetColorSpace | Quality |      Mean |
|---------- |----------------- |-------- |----------:|
| Benchmark |        Luminance |      75 |  6.923 ms |
| Benchmark |              Rgb |      75 | 12.116 ms |
| Benchmark |    YCbCrRatio420 |      75 |  6.484 ms |
| Benchmark |    YCbCrRatio444 |      75 |  8.263 ms |
| Benchmark |        Luminance |      90 |  7.046 ms |
| Benchmark |              Rgb |      90 | 13.349 ms |
| Benchmark |    YCbCrRatio420 |      90 |  6.714 ms |
| Benchmark |    YCbCrRatio444 |      90 |  9.198 ms |
| Benchmark |        Luminance |     100 |  9.057 ms |
| Benchmark |              Rgb |     100 | 19.099 ms |
| Benchmark |    YCbCrRatio420 |     100 | 10.194 ms |
| Benchmark |    YCbCrRatio444 |     100 | 15.372 ms |


PR

|    Method | TargetColorSpace | Quality |      Mean |
|---------- |----------------- |-------- |----------:|
| Benchmark |        Luminance |      75 |  4.618 ms |
| Benchmark |              Rgb |      75 | 12.543 ms |
| Benchmark |    YCbCrRatio420 |      75 |  6.639 ms |
| Benchmark |    YCbCrRatio444 |      75 |  8.590 ms |
| Benchmark |        Luminance |      90 |  5.456 ms |
| Benchmark |              Rgb |      90 | 13.447 ms |
| Benchmark |    YCbCrRatio420 |      90 |  7.218 ms |
| Benchmark |    YCbCrRatio444 |      90 |  9.150 ms |
| Benchmark |        Luminance |     100 |  6.731 ms |
| Benchmark |              Rgb |     100 | 19.831 ms |
| Benchmark |    YCbCrRatio420 |     100 | 10.541 ms |
| Benchmark |    YCbCrRatio444 |     100 | 15.345 ms |

br3aker commented:
Need to write a ton of tests now....

br3aker commented on src/ImageSharp/Formats/Jpeg/Components/Encoder/ComponentProcessor.cs:
```diff
@@ -0,0 +1,233 @@
+                MultiplyToAverage(sourceRow, averageMultiplier);
+
+                // copy to the first 8 slots
+                sourceRow.Slice(0, packedWidth).CopyTo(this.ColorBuffer.DangerousGetRowSpan(i / factors.Height));
```
This is one of the reasons 420 subsampling encoding got slower - we abosolutely need to move ready-to-encode color strides to the first 8 indices. This potentially can be resolved at the TPixel -> RGB conversion step - we can rearrange color strides before doing subsampling but it's a very tricky thing to do, I've decided to make this PR as small as possible.

gfoidl commented on src/ImageSharp/Formats/Jpeg/Components/Block8x8F.ScaledCopy.cs:
```diff
@@ -0,0 +1,160 @@
+                        for (int j = 0; j < horizontalScale; j++)
+                        {
+                            // area[xx + j, yy + i] = value;
+                            Unsafe.Add(ref areaOrigin, baseIdx + j) = value;
```
```suggestion
                            Unsafe.Add(ref areaOrigin, (nint)(uint)(baseIdx + j)) = value;
```
Or change the types to `nint` (or `nuint`) to avoid the casts here.

gfoidl reviewed (commented):
Half way viewed (have to leave now, will continue later).

gfoidl commented on src/ImageSharp/Formats/Jpeg/Components/Encoder/Component.cs:
```diff
@@ -0,0 +1,113 @@
+
+            this.HorizontalSamplingFactor = horizontalFactor;
+            this.VerticalSamplingFactor = verticalFactor;
+            this.SamplingFactors = new Size(this.HorizontalSamplingFactor, this.VerticalSamplingFactor);
```
```suggestion
            this.SamplingFactors = new Size(horizontalFactor, verticalFactor);
```
I guess this produces better code, especially when the JIT has some work with inlining, etc. (to not reach some CSE limits and so on).

gfoidl commented on src/ImageSharp/Formats/Jpeg/Components/Encoder/Component.cs:
```diff
@@ -0,0 +1,113 @@
+        public void Init(JpegFrame frame, int maxSubFactorH, int maxSubFactorV)
+        {
+            this.WidthInBlocks = (int)MathF.Ceiling(
+                MathF.Ceiling(frame.PixelWidth / 8F) * this.HorizontalSamplingFactor / maxSubFactorH);
```
```suggestion
                ((uint)frame.PixelWidth + 7) / 8) * this.HorizontalSamplingFactor / maxSubFactorH);
```
Produces (only for the MathF.Ceiling) part [way better code](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIGYACMhgYQYG8aHunykGAlgDsMDAGIQIACmGiADgIQwANgHUBAEwwALAJQMAvAD4GMkboCy2HWIB0LGAOXCA5lIVK1mnQwD0DAA4AM10Abi4eYj5BEQYAIWwoM3lFFXUtPUMTZN0pPIBXWV0PNO9tBgBqBgB2fX8AsJoAXyA==)

gfoidl commented on src/ImageSharp/Formats/Jpeg/Components/Encoder/Component.cs:
```diff
@@ -0,0 +1,113 @@
+                MathF.Ceiling(frame.PixelWidth / 8F) * this.HorizontalSamplingFactor / maxSubFactorH);
+
+            this.HeightInBlocks = (int)MathF.Ceiling(
+                MathF.Ceiling(frame.PixelHeight / 8F) * this.VerticalSamplingFactor / maxSubFactorV);
```
```suggestion
                ((uint)frame.PixelHeight + 7) / 8) * this.VerticalSamplingFactor / maxSubFactorV);
```

gfoidl commented on src/ImageSharp/Formats/Jpeg/Components/Encoder/ComponentProcessor.cs:
```diff
@@ -0,0 +1,233 @@
+                    int length = target.Length / Vector256<float>.Count;
+                    for (int i = 0; i < haddIterationsCount; i++)
+                    {
+                        length /= 2;
```
```suggestion
                        length >>= 1;    // bit-hack for / 2
```
or make length an unsiged type (`uint` or `nuint`).

gfoidl commented on src/ImageSharp/Formats/Jpeg/Components/Encoder/ComponentProcessor.cs:
```diff
@@ -0,0 +1,233 @@
+                    for (int i = 0; i < haddIterationsCount; i++)
+                    {
+                        length /= 2;
+                        for (int j = 0; j < length; j++)
```
```suggestion
                        for (nuint j = 0; j < length; j++)
```
also for `indexLeft` and `indexRight`.

gfoidl commented on src/ImageSharp/Formats/Jpeg/Components/Encoder/HuffmanScanEncoder.cs:
```diff
@@ -128,57 +129,67 @@ private bool IsStreamFlushNeeded
+                Span<Block8x8> blockSpan = component.SpectralBlocks.DangerousGetRowSpan(y: 0);
+                ref Block8x8 blockRef = ref MemoryMarshal.GetReference(blockSpan);
+
+                for (int k = 0; k < w; k++)
```
```suggestion
                for (nint k = 0; k < w; k++)
```

gfoidl commented on src/ImageSharp/Formats/Jpeg/Components/Encoder/HuffmanScanEncoder.cs:
```diff
@@ -191,65 +202,33 @@ public void Encode444<TPixel>(Image<TPixel> pixels, ref Block8x8F luminanceQuant
-                        prevDCCr,
-                        ref pixelConverter.Cr,
-                        ref chrominanceQuantTable);
+                for (int k = 0; k < w; k++)
```
```suggestion
                for (nint k = 0; k < w; k++)
```

gfoidl commented on src/ImageSharp/Formats/Jpeg/Components/Encoder/HuffmanScanEncoder.cs:
```diff
@@ -262,43 +241,62 @@ public void Encode420<TPixel>(Image<TPixel> pixels, ref Block8x8F luminanceQuant
+
+                            for (int x = 0; x < h; x++)
+                            {
+                                int blockCol = (mcuCol * h) + x;
```
JIT won't hoist `mcuCol * h` and this product is loop-invariant, so hoist it manually above the two for-loops.
I'd also make the type of `blockCol` `nint` for the `Unsafe.Add` below to avoid the sign-extending move.

gfoidl commented on src/ImageSharp/Formats/Jpeg/Components/Encoder/HuffmanScanEncoder.cs:
```diff
@@ -310,54 +308,59 @@ public void EncodeGrayscale<TPixel>(Image<TPixel> pixels, ref Block8x8F luminanc
+                converter.ConvertStrideBaseline();
+
+                // Encode spectral to binary
+                for (int i = 0; i < mcusPerLine; i++)
```
```suggestion
                for (nint i = 0; i < (nint)(uint)mcusPerLine; i++)
```
I guess the casts are necessary for the JIT to keep that expression in a register.

gfoidl commented on src/ImageSharp/PixelFormats/PixelOperations{TPixel}.cs:
```diff
@@ -198,6 +198,37 @@ public partial class PixelOperations<TPixel>
+            ref float g = ref MemoryMarshal.GetReference(greenChannel);
+            ref float b = ref MemoryMarshal.GetReference(blueChannel);
+            ref TPixel src = ref MemoryMarshal.GetReference(source);
+            for (int i = 0; i < count; i++)
```
```suggestion
            for (nint i = 0; i < (nint)(uint)count; i++)
```

gfoidl reviewed (commented):
Finished my review -- didn't look at test code (due to missing knowledge there).

Looks good 👍🏻

br3aker commented:
**Update**

Added YccK encoding so this PR completely resolves https://github.com/SixLabors/ImageSharp/issues/808.

br3aker commented:
Hmm, this failed tests passes on my local machine with .net6 and fails only on macos on .net7. Wondering if this is the same problem as in https://github.com/SixLabors/ImageSharp/issues/2117.

brianpopow commented:
> Hmm, this failed test passes on my local machine with .net6 and fails only on macos on .net7. Wondering if this is the same problem as in #2117.

I have not seen issue #2117 on mac yet, only on windows so far, but it could be related.

I will restart the CI to see, if it was a fluke or always happens.

br3aker commented:
> I will restart the CI to see, if it was a fluke or always happens.

Looks like it's not a fluke. I'll start writing tests this week for this PR and try to add debug checks for memory access, maybe it's related to some out of bounds index access.

br3aker commented:
Finally had a moment to finish this:
1. All colors supported for encoding
2. Added tests and removed obsolete benchmark tests - we have proper benchmarks for that 
3. Restored `ProfilingSandbox.cs`

@JimBobSquarePants @brianpopow

br3aker commented:
Looks like `Skew` tests fails on net7 is still a thing. Everything else should be working just fine. Added tests for all new output color types with a very precise tolerance.

JimBobSquarePants commented:
Very exciting! Will review this asap.

br3aker commented:
Looks like I've messed up linked issues, here's the complete list of what's resolved by this PR:
1. https://github.com/SixLabors/ImageSharp/issues/1476 - all color converions now supports scalar/vector/avx paths
2. https://github.com/SixLabors/ImageSharp/issues/808 - all output color types are supported
3. https://github.com/SixLabors/ImageSharp/issues/1713 - only required quantization table(s) is(are) written
4. https://github.com/SixLabors/ImageSharp/issues/1238 - CMYK input jpeg now should be resaved to CMYK output jpeg if not explicitely overriden in encoder settings

JimBobSquarePants reviewed (approved):
This all looks awesome. Thanks again for such a fantastic addition to the library!
